File: 383-ransom-note/383-ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        ransomNoteDict = Counter(ransomNote)
        magazineDict = Counter(magazine)
        for k in ransomNoteDict:
            magazineDict[k] = magazineDict.get(k, 0) - ransomNoteDict[k]
            if magazineDict[k] < 0:
                return False
        return True

File: 383-ransom-note/383-ransom-note.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function(ransomNote, magazine) {
    
    if (ransomNote === null || magazine === null || magazine.length < ransomNote.length){
        return false
    }
    
    const charMap = new Map()
    
    for (let char of magazine){
        charMap.set(char, charMap.get(char) ? charMap.get(char) + 1 : 1)
    }
    
    for (let char of ransomNote){
        if( !charMap.has(char) || charMap.get(char) === 0 ) {
            return false
        } else {
            charMap.set(char, charMap.get(char) ? charMap.get(char) - 1 : 0)
        }
        
    }
    
    return true

};